### PR TITLE
MCO-565: MCO-568: MCO-659: MCO-660 On-cluster build opt-in function, building machine-os-builder stub, RBAC and service acct inclusions. Deletes deployment rather than scale down to 0 without label

### PR DESCRIPTION
**Pull request by dkhater-redhat**
**- What I did**
changed so that machine-os-builder deployment is removed when "layering-enabled" label is removed, rather than scale the deployment down to 0.

**- How to verify it**
`~/MCO/on-cluster-build on ☁️  (us-east-1) on ☁️  aos-serviceaccount@openshift-gce-devel.iam.gserviceaccount.com 
❯ oc describe mcp | grep "layering-enabled"
Labels:       machineconfiguration.openshift.io/layering-enabled=
          f:machineconfiguration.openshift.io/layering-enabled:

~/MCO/on-cluster-build on ☁️  (us-east-1) on ☁️  aos-serviceaccount@openshift-gce-devel.iam.gserviceaccount.com took 5s 
❯ oc get pods
NAME                                         READY   STATUS      RESTARTS   AGE
machine-config-controller-5d695f9d96-qtzdh   2/2     Running     0          3m45s
machine-config-daemon-f2p4d                  2/2     Running     0          3m27s
machine-config-daemon-hh2vg                  2/2     Running     0          3m35s
machine-config-daemon-mt8l6                  2/2     Running     0          3m51s
machine-config-daemon-n5qht                  2/2     Running     0          3m43s
machine-config-daemon-nntrk                  2/2     Running     0          3m25s
machine-config-daemon-t6895                  2/2     Running     0          3m24s
machine-config-operator-5cb58c56bb-fhk7d     2/2     Running     0          3m45s
machine-config-server-7669w                  1/1     Running     0          3m52s
machine-config-server-tscbw                  1/1     Running     0          3m45s
machine-config-server-vzqtd                  1/1     Running     0          3m37s
machine-os-builder-67f76d7f49-cdmfz          1/1     Running     0          3m19s
mco-image-build-build                        0/1     Completed   0          11m

~/MCO/on-cluster-build on ☁️  (us-east-1) on ☁️  aos-serviceaccount@openshift-gce-devel.iam.gserviceaccount.com 
❯ oc get deployment
NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
machine-config-controller   1/1     1            1           6h53m
machine-config-operator     1/1     1            1           6h58m
machine-os-builder          1/1     1            1           3m24s

~/MCO/on-cluster-build on ☁️  (us-east-1) on ☁️  aos-serviceaccount@openshift-gce-devel.iam.gserviceaccount.com 
❯ oc label mcp/on-cluster-build 'machineconfiguration.openshift.io/layering-enabled-'
machineconfigpool.machineconfiguration.openshift.io/on-cluster-build unlabeled

~/MCO/on-cluster-build on ☁️  (us-east-1) on ☁️  aos-serviceaccount@openshift-gce-devel.iam.gserviceaccount.com took 5s 
❯ oc get pods
NAME                                         READY   STATUS      RESTARTS   AGE
machine-config-controller-5d695f9d96-qtzdh   2/2     Running     0          4m21s
machine-config-daemon-f2p4d                  2/2     Running     0          4m3s
machine-config-daemon-hh2vg                  2/2     Running     0          4m11s
machine-config-daemon-mt8l6                  2/2     Running     0          4m27s
machine-config-daemon-n5qht                  2/2     Running     0          4m19s
machine-config-daemon-nntrk                  2/2     Running     0          4m1s
machine-config-daemon-t6895                  2/2     Running     0          4m
machine-config-operator-5cb58c56bb-fhk7d     2/2     Running     0          4m21s
machine-config-server-7669w                  1/1     Running     0          4m28s
machine-config-server-tscbw                  1/1     Running     0          4m21s
machine-config-server-vzqtd                  1/1     Running     0          4m13s
mco-image-build-build                        0/1     Completed   0          12m
^C%                                                                                                                                                       
~/MCO/on-cluster-build on ☁️  (us-east-1) on ☁️  aos-serviceaccount@openshift-gce-devel.iam.gserviceaccount.com took 8s 
❯ oc get deployment
NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
machine-config-controller   1/1     1            1           6h54m
machine-config-operator     1/1     1            1           6h59m`

**- Description for the changelog**

### CODE REVIEW

File: test/helpers/utils.go [hunk 1]
```diff
@@ -329,6 +329,65 @@ func WaitForPausedConfig(t *testing.T, cs *framework.ClientSet, pool string) err
...
+		// If we reached here, it means no pod with the given prefix exists, so we return true
+		return true, nil
+	})
+}
+
+// CheckDeploymentExists checks if a deployment with the given name in the given namespace exists.
+func CheckDeploymentExists(cs *framework.ClientSet, name, namespace string) (bool, error) {
+	ctx := context.TODO()
+	_, err := cs.Deployments(namespace).Get(ctx, name, metav1.GetOptions{})
+	// If an error occurs and it's a NotFound error, return false, nil
```
Comment (cheesesashimi) [inline test/helpers/utils.go:379]: suggestion: We can make the detection logic a bit easier to follow / read, e.g.:

```golang
if err == nil {
	return true, nil
}

if errors.IsNotFound(err) {
	return false, nil
}

return false, err
```

**Review by cheesesashimi — CHANGES_REQUESTED** [review]
Overall this looks good! I have some ideas and suggestions to help improve this.

**Comment by sergiordlr** [comment]
No builder pod deployed

```
$  oc get pods -n openshift-machine-config-operator 
NAME                                         READY   STATUS    RESTARTS   AGE
machine-config-controller-74c87d6687-rl7zq   2/2     Running   0          18m
machine-config-daemon-66jgb                  2/2     Running   0          18m
machine-config-daemon-h6jkd                  2/2     Running   0          18m
machine-config-daemon-j9wvd                  2/2     Running   0          10m
machine-config-daemon-mt4cs                  2/2     Running   0          9m8s
machine-config-daemon-ngmj6                  2/2     Running   0          10m
machine-config-daemon-pxwjt                  2/2     Running   0          18m
machine-config-operator-6457d8bd78-65m2x     2/2     Running   0          21m
machine-config-server-cnlrp                  1/1     Running   0          17m
machine-config-server-gpb8h                  1/1     Running   0          17m
machine-config-server-nkl7m                  1/1     Running   0          17m
```

Verify synced resources:

```

$ oc get clusterrole  |grep builder
machine-os-builder                                                          2023-08-09T10:17:56Z
machine-os-builder-events                                                   2023-08-09T10:17:56Z


$  oc get ClusterRoleBinding |grep builder
machine-os-builder                                                          ClusterRole/machine-os-builder                                                          17m
machine-os-builder-anyuid                                                   ClusterRole/system:openshift:scc:anyuid   

# Note that the machine-os-builder deployment does not exist
$ oc get deployment -n openshift-machine-config-operator 
NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
machine-config-controller   1/1     1            1           19m
machine-config-operator     1/1     1            1           28m

$ oc get rolebindings -n default |grep event
machine-config-controller-events   ClusterRole/machine-config-controller-events   19m
machine-config-daemon-events       ClusterRole/machine-config-daemon-events       20m
machine-os-builder-events          ClusterRole/machine-os-builder-events          18m


$ oc get rolebindings -n openshift-machine-config-operator |grep event
machine-config-controller-events   ClusterRole/machine-config-controller-events   20m
machine-config-daemon-events       ClusterRole/machine-config-daemon-events       20m
machine-os-builder-events          ClusterRole/machine-os-builder-events          19m

$ oc get sa -n openshift-machine-config-operator machine-os-builder
NAME                 SECRETS   AGE
machine-os-builder   1         19m

$  oc rsh -n openshift-machine-config-operator daemonset/machine-config-daemon ls -larth /usr/bin |grep machine
Defaulted container "machine-config-daemon" out of: machine-config-daemon, oauth-proxy
-rwxr-xr-x. 1 root root  25K Apr 27 12:11  systemd-machine-id-setup
-rwxr-xr-x. 1 root root 1.8M Aug  9 09:55  machine-os-builder   <------ BUILDER IS INCLUDED IN THE IMAGE
-rwxr-xr-x. 1 root root  44M Aug  9 09:55  machine-config-server
-rwxr-xr-x. 1 root root  51M Aug  9 09:55  machine-config-operator
-rwxr-xr-x. 1 root root  55M Aug  9 09:55  machine-config-daemon
-rwxr-xr-x. 1 root root  80M Aug  9 09:55  machine-config-controller
-rwxr-xr-x. 1 root root  55M Aug  9 09:56  machine-config-daemon.rhel9
```




Verify builder deployment


1) When we label the worker MCP a machine-os-builder deployment is created and scaled to 1 replicas
```
$ oc label mcp worker machineconfiguration.openshift.io/layering-enabled=
machineconfigpool.machineconfiguration.openshift.io/worker labeled

$ oc get deployment -n openshift-machine-config-operator 
NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
machine-config-controller   1/1     1            1           22m
machine-config-operator     1/1     1            1           31m
machine-os-builder          1/1     1            1           17s

$ oc get pods -n openshift-machine-config-operator  |grep builder
machine-os-builder-7dc56d9f77-zcnmv          1/1     Running   0          87s

$ oc logs -n openshift-machine-config-operator  machine-os-builder-7dc56d9f77-zcnmv
Hello, World!
```


2) When we unlabel the worker MCP the machine-os-builder is removed

```
]$ oc label mcp worker machineconfiguration.openshift.io/layering-enabled-
machineconfigpool.machineconfiguration.openshift.io/worker unlabeled

$ oc get deployment
NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
machine-config-controller   1/1     1            1           29m
machine-config-operator     1/1     1            1           38m
$ oc get replicaset
NAME                                   DESIRED   CURRENT   READY   AGE
machine-config-controller-74c87d6687   1         1         1       29m
machine-config-operator-6457d8bd78     1         1         1       38m

```

3. If we use master MCP we get the same behavior

4. If we label both MCPs at the same time, we get a machine-os-builder deployment with 1 replica. This new deployment is not removed until the labels are removed from both MCPs.

5. With a custom infra MCP the behavior is the same

6. If we label a MCP and once the machine-os-builder deployment is created we manually scale the deployment to 0 replicas, then the deployment remains with 0 replicas and it is never restored to the original 1 replicas.

```
$ oc scale deployment machine-os-builder --replicas 0
deployment.apps/machine-os-builder scaled
$ oc get deployment
NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
machine-config-controller   1/1     1            1           40m
machine-config-operator     1/1     1            1           49m
machine-os-builder          0/0     0            0           2m49s
```

7. In SNO the behavior is working as expected. Only one thing to note, when we label the worker pool (with 0 nodes) the machine-os-builder is created too. It doesn't look like a problem, though.


Please, could you confirm that the behavior in step 6 is a desired behavior? if we manually scale the machine-os-builder deployment to another value, should the replicas be restored to its original value or not?

**Comment by dkhater-redhat** [comment]
```
❯ oc get deployment
NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
machine-config-controller   1/1     1            1           6h40m
machine-config-operator     1/1     1            1           6h45m
machine-os-builder          1/1     1            1           12s

~/MCO/on-cluster-build on ☁️  (us-east-1) on ☁️  aos-serviceaccount@openshift-gce-devel.iam.gserviceaccount.com 
❯ oc scale deployment machine-os-builder --replicas 0                                
deployment.apps/machine-os-builder scaled

~/MCO/on-cluster-build on ☁️  (us-east-1) on ☁️  aos-serviceaccount@openshift-gce-devel.iam.gserviceaccount.com 
❯ oc get deployment --watch
NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
machine-config-controller   1/1     1            1           6h40m
machine-config-operator     1/1     1            1           6h45m
machine-os-builder          0/0     0            0           24s
machine-os-builder          0/1     0            0           34s
machine-os-builder          0/1     0            0           34s
machine-os-builder          0/1     0            0           35s
machine-os-builder          0/1     1            0           35s
machine-os-builder          1/1     1            1           35s
^C%                                                                                                                                                         
~/MCO/on-cluster-build on ☁️  (us-east-1) on ☁️  aos-serviceaccount@openshift-gce-devel.iam.gserviceaccount.com took 19s 
❯ oc get deployment
NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
machine-config-controller   1/1     1            1           6h41m
machine-config-operator     1/1     1            1           6h45m
machine-os-builder          1/1     1            1           46s
```

**Comment by sergiordlr** [comment]
After the latest changes if we manually modify the number of replicas for the machine-os-builder deployment the replicas are automatically restored to 1. (tested with replicas=0 and replicas=5)

Everything seems to be working fine, thank you very much @dkhater-redhat !!

We can add the qe-approved label

/label qe-approved

**Comment by jkyros** [comment]
I'm sure it's fine, but just in case. 
/payload-job periodic-ci-openshift-release-master-nightly-4.14-e2e-aws-ovn-single-node

**Comment by sinnykumari** [comment]
Re-running periodic job in case it was a flake
/payload-job periodic-ci-openshift-release-master-nightly-4.14-e2e-aws-ovn-single-node

**Comment by dkhater-redhat** [comment]
looking at the MCO Pod logs, everything seems to be good. Failure seems to be unrelated to the changes made in this PR

**Comment by cheesesashimi** [comment]
I can confirm that the failures identified in https://pr-payload-tests.ci.openshift.org/runs/ci/22e847d0-3a8a-11ee-9375-c0f0d7e9391c-0 are unrelated to this PR. Overall, this looks good enough to merge.

/lgtm
/approval

**Pull request merged**